File: eduarda/pset6/gincana.py
if __name__ == "__main__":
    n, m = map(int, input().split())

    vertices = [[] for _ in range(n)]    
    for _ in range(m):
        i, j = map(int, input().split())
        i -= 1
        j -= 1        
        vertices[i].append(j)
        vertices[j].append(i)
   
    painted = [False for _ in range(n)]
    groups = 0

    for i in range(n):
        if painted[i] == False:
            groups += 1
            pilha = [i]
            painted[i] = True
            while len(pilha) > 0:
                a = pilha.pop()
                painted[a] = True

                amigos = vertices[a]
               
                for ai in amigos:
                    if painted[ai] == False:
                        pilha.append(ai)
       
    print(groups)

File: eduarda/pset6/gincana.cpp
#include <bits/stdc++.h>
using namespace std;


int main(){
    int n, m;
    cin >> n >> m;
    vector<vector<int>> vertices(n);
    while(m--){
        int a, b;
        cin >> a >> b;
        a -= 1;
        b -= 1;
        vertices[a].push_back(b);
        vertices[b].push_back(a);
    }  

    vector<bool> visited(n, false);
    int groups = 0;
    stack<int> stackers;

    for(int i = 0; i < n; i++){
        if(!visited[i]){
            groups += 1;
            visited[i] = true;
            stackers.push(i);
            while(!stackers.empty()){
                int a = stackers.top();
                visited[a] = true;
                stackers.pop();
                vector<int> friends = vertices[a];
                for(auto fr : friends){
                    if(!visited[fr]){
                        stackers.push(fr);
                    }
                }
            }
        }
    }

    cout<< groups;
}
